Add MirroredMessageUsersTest.

showell opened the pull request:
See #1006 

Once this hits master, I'll hit some of the more fiddly cases, like ist.mit.edu users and domains that end in "irc" (I think there's a very minor bug in the code that would impact a realm like blirc.com).

FYI I may have sporadic Internet access in the next couple days, so for any simple changes, please just make them on my behalf.

smarx commented:
*Automated message from Dropbox CLA bot*

@showell, it looks like you've already signed the Dropbox CLA. Thanks!

showell commented:
One observation about the code under test here is that we could probably have the caller grab the relevant data out of the request object (e.g. request.POST['sender'] and request.client.name), so that it's a little easier to test.   (And maybe we'd make it an action?)  The caller already validates that 'sender' is in request.POST, so we are doing double validation here, and the caller already does `client = request.client`, so we are duplicating some object dereferencing.

timabbott commented on zerver/tests/test_messages.py:
```diff
@@ -900,6 +903,152 @@ def test_propagate_all_topics(self):
+        realm_users = UserProfile.objects.filter(realm=sender.realm)
+        realm_emails = {user.email for user in realm_users}
+        self.assertIn('alice@mit.edu', realm_emails)
+        self.assertIn('bob_the_new_user@mit.edu', realm_emails)
```
Would check `is_mirror_dummy` is set for the new bob account

showell replied:
Amended, and fixed a `class Foo(object)` lint thing.  Signing off for now.

timabbott replied:
Cool.  I'll add this check in the other two tests as well.

timabbott commented:
I added a second zephyr_mirror test for the case where the sender is to be created, addressed my small comment above, and merged this.

That said, I'm actually not sure that this test suite wouldn't be even better if we replaced it with a series of tests for the `send_message` API calls that exercise these same cases (rather than using mock Request objects), since I don't think it'd cost much more in terms of performance, the test code would be shorter, and it'd be a bit more end-to-end (and thus robust to e.g. our refactoring the APIs as you suggest above, which I think would be an improvement)).

showell commented:
The one existing test I saw was more end to end and seemed to have some complicated side effects that required mocking, so I was a bit timid.  I definitely support doing an another iteration on this, although I am going breadth first on the coverage for subsystems that are mostly outside my expertise.  Thx for fixing/merging this!

Sent from my iPhone

> On Jul 26, 2016, at 8:05 PM, Tim Abbott <notifications@github.com> wrote:
> 
> I added a second zephyr_mirror test for the case where the sender is to be created, addressed my small comment above, and merged this.
> 
> That said, I'm actually not sure that this test suite wouldn't be even better if we replaced it with a series of tests for the send_message API calls that exercise these same cases (rather than using mock Request objects), since I don't think it'd cost much more in terms of performance, the test code would be shorter, and it'd be a bit more end-to-end (and thus robust to e.g. our refactoring the APIs as you suggest above, which I think would be an improvement)).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
